### ISSUE TITLE
[v3] Implement new Concept Exercise: `with`

### ISSUE THREAD

**Issue by angelikatyborska** [exercism/elixir#560]
## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md)

## Goal

The goal of this exercise is to teach the student how the `with` special form works and when it is appropriate to use (e.g. to avoid nested `case`s)

## Learning objectives

- Know how to use `with` to chain matching clauses 
- Know how to use the `else` block in `with` to match on different kinds of errors
- Know when _not_ to use `with`

## Out of scope

?

## Concepts

- `with`

## Prerequisites

- `errors` (the `with` special form is most commonly used with error/ok-tuples and the exercise should teach that)
- `pattern-matching`
- other, depending on the story

## Resources to refer to

- https://hexdocs.pm/elixir/Kernel.SpecialForms.html#with/1
- https://elixir-lang.org/getting-started/mix-otp/docs-tests-and-with.html#with
- http://learningelixir.joekain.com/learning-elixir-with/
- https://openmymind.net/Elixirs-With-Statement/

## Analyzer

It will be necessary to use the analyzer to check that the `with` form was indeed used, and that no alternatives (e.g. `case`s) were used.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

**Comment by ThiagosLima** [comment]
Hi, I am still learning Elixir and want to help with some issues in the future. But for now, I have just a comment. 

This Analyzer part that assure students use the correct function could be added in some exercises. Like **Strain** exercise which it is not supposed to use `filter` and `reject`. And the **Accumulate** exercise that it is note supposed to use `collect`, `map` or `fmap`.

I don't know if it is already done in the v3. By the way, nice work with the Elixir track. It is really good!

**Comment by neenjaw** [comment]
@ThiagosLima thanks for the comment!! Glad to have you here! Yes, that is the eventual goal to add this functionality to practice exercises at some point too. :)

If you want to open an issue in our analyzer repo. Would help us to keep track of that: https://github.com/exercism/elixir-analyzer

**Comment by ThiagosLima** [comment]
Thanks! @neenjaw that is nice! I didn't know about the analyzer repo haha! Sorry. I will add the issues for the exercises that I saw.

**Comment by jersearls** [comment]
Howdy! I'm interested in building this exercise out. Is anyone actively working on it? @angelikatyborska

**Comment by angelikatyborska** [comment]
@jersearls We might have one other person interested but they're only starting to learn how to build concept exercises, so it's unclear. I consider this issue available until a get a draft PR from anyone with at least an example solution (`<exercise-name>/.meta/exemplar.ex`) which is usually the first step for building those exercises (maybe together with the tests).

**Comment by papey** [comment]
> @jersearls We might have one other person interested but they're only starting to learn how to build concept exercises, so it's unclear. I consider this issue available until a get a draft PR from anyone with at least an example solution (`<exercise-name>/.meta/exemplar.ex`) which is usually the first step for building those exercises (maybe together with the tests).

yay, it's me :smile: 

@jersearls feel free to start working on it since I have to learn a bunch of things in the process you might be faster than me.

**Comment by jiegillet** [comment]
Closed via #954.
